currencyrate/currencyrate.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+from pyln.client import Plugin
+from collections import namedtuple
+from pyln.client import Millisatoshi
+from cachetools import cached, TTLCache
+import requests
+import statistics
+import time
+
+plugin = Plugin()
+
+Source = namedtuple('Source', ['name', 'urlformat', 'replymembers'])
+
+sources = [
+    # e.g. {"GBP": {"volume_btc": "24.36647424", "rates": {"last": "13667.63"}, "avg_1h": "13786.86", "avg_6h": "13723.65", "avg_12h": "13680.23", "avg_24h": "13739.56"}, "USD": {"volume_btc": "27.97517017", "rates": {"last": "18204.21"}, "avg_1h": "19349.46", "avg_6h": "18621.72", "avg_12h": "18642.28", "avg_24h": "18698.94"}
+    # ...
+    # "GTQ": {"volume_btc": "0.03756101", "rates": {"last": "148505.46"}, "avg_1h": "148505.46", "avg_6h": "162463.69", "avg_12h": "162003.10", "avg_24h": "162003.10"}, "DKK": {"volume_btc": "0.00339923", "rates": {"last": "139737.53"}, "avg_12h": "139737.53", "avg_24h": "139737.53"}, "HTG": {"volume_btc": "0.00024758", "rates": {"last": "2019549.24"}, "avg_6h": "2019549.24", "avg_12h": "2019549.24", "avg_24h": "2019549.24"}, "NAD": {"volume_btc": "0.00722222", "rates": {"last": "360000.11"}, "avg_12h": "360000.11", "avg_24h": "360000.11"}}
+    Source('localbitcoins',
+           'https://localbitcoins.com/bitcoinaverage/ticker-all-currencies/',
+           ['{currency}', "avg_6h"]),
+    # e.g. {"high": "18502.56", "last": "17970.41", "timestamp": "1607650787", "bid": "17961.87", "vwap": "18223.42", "volume": "7055.63066541", "low": "17815.92", "ask": "17970.41", "open": "18250.30"}
+    Source('bitstamp',
+           'https://www.bitstamp.net/api/v2/ticker/btc{currency_lc}/',
+           ['last']),
+    # e.g. {"bitcoin":{"usd":17885.84}}
+    Source('coingecko',
+           'https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies={currency_lc}',
+           ['bitcoin', '{currency_lc}']),
+]
+
+
+def get_currencyrate(plugin, currency, req_template, response_members):
+    # NOTE: Bitstamp has a DNS/Proxy issues that can return 404
+    # Workaround: retry up to 5 times with a delay
+    currency_lc = currency.lower()
+    url = req_template.format(currency_lc=currency_lc, currency=currency)
+    for _ in range(5):
+        r = requests.get(url, proxies=plugin.proxies)
+        if r.status_code != 200:
+            time.sleep(1)
+            continue
+        break
+
+    if r.status_code != 200:
+        plugin.log(level='info', message='{}: bad response {}'.format(url, r.status_code))
+        return None
+
+    json = r.json()
+    for m in response_members:
+        expanded = m.format(currency_lc=currency_lc, currency=currency)
+        if expanded not in json:
+            plugin.log(level='debug', message='{}: {} not in {}'.format(url, expanded, json))
+            return None
+        json = json[expanded]
+
+    try:
+        return Millisatoshi(int(10**11 / float(json)))
```

Comment:
Technically this is not `Millisatoshi` it's `Millisatoshi / currency` which we don't have a type for. We could keep it unitless or introduce a `Rate` class.

Reply:
No, it's the number of millisatoshi equivalent to 1 currency unit?

Reply:
Hehe, that's just rephrasing it, I could just as well say it's number of millisatoshi units / currency, which would come out as the inverse of the currency. Currencies aren't unitless amounts.

Anyway, splitting hairs here, we can move on to more important things :-)